src/main/java/apoc/algo/WeaklyConnectedComponents.java

```diff
@@ -0,0 +1,105 @@
+package apoc.algo;
+
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.stream.Stream;
+
+import org.neo4j.graphdb.Direction;
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Node;
+import org.neo4j.graphdb.Relationship;
+import org.neo4j.graphdb.RelationshipType;
+import org.neo4j.graphdb.ResourceIterator;
+import org.neo4j.logging.Log;
+import org.neo4j.procedure.Context;
+import org.neo4j.procedure.Procedure;
+
+import com.carrotsearch.hppc.LongHashSet;
+import com.carrotsearch.hppc.cursors.LongCursor;
+
+import apoc.Description;
+import apoc.result.LongResult;
+
+public class WeaklyConnectedComponents {
+
+	@Context
+	public GraphDatabaseService db;
+
+	@Context
+	public Log log;
+
+	@Procedure("apoc.algo.wcc")
+	@Description("CALL apoc.algo.wcc() YIELD number of weakly connected components")
+	public Stream<LongResult> wcc() {
+		try {
+			long componentID = 0;
+			ResourceIterator<Node> nodes = db.getAllNodes().iterator();
+			LongHashSet allNodes = new LongHashSet();
+			while (nodes.hasNext()) {
+				Node node = nodes.next();
+				if (node.getDegree() == 0) {
+					componentID++;
+				} else {
+					allNodes.add(node.getId());
+				}
+			}
+			nodes.close();
+
+			Iterator<LongCursor> it = allNodes.iterator();
+			while (it.hasNext()) {
+				// Every node has to be marked as (part of) a component
+				try {
+					Long n = it.next().value;
+					LongHashSet reachableIDs = go(db.getNodeById(n), Direction.BOTH);
+					allNodes.removeAll(reachableIDs);
+					componentID++;
+
+				} catch (NoSuchElementException e) {
+					break;
+				}
+				it = allNodes.iterator();
+			}
+			return Stream.of(new LongResult(componentID));
```

Comment:
Alternatively it could directly set the component-id as a property (configurable) on those nodes. then it would need `@PerformsWrites` annotation